tls/s2n_send.c

```diff
@@ -103,6 +103,37 @@ int s2n_flush(struct s2n_connection *conn, s2n_blocked_status *blocked)
     return 0;
 }
 
+S2N_RESULT s2n_sendv_with_offset_total_size(const struct iovec *bufs, ssize_t count,
+        ssize_t offs, ssize_t *total_size_out)
+{
+    RESULT_ENSURE_REF(total_size_out);
+    if (count) {
+        RESULT_ENSURE_REF(bufs);
+    }
+
+    size_t total_size = 0;
+    for (ssize_t i = 0; i < count; i++) {
+        size_t iov_len = bufs[i].iov_len;
+        /* Account for any offset */
+        if (offs) {
+            size_t offs_consumed = MIN((size_t) offs, iov_len);
+            iov_len -= offs_consumed;
+            offs -= offs_consumed;
+        }
+        RESULT_ENSURE(S2N_ADD_IS_OVERFLOW_SAFE(total_size, iov_len, SSIZE_MAX),
+                S2N_ERR_INVALID_ARGUMENT);
```

Comment:
nit: Is there a difference between SSIZE_MAX and SIZE_MAX? Is SIZE_MAX more correct here?

Reply:
SSIZE_MAX is the eventual limit we want to enforce and works here because of how I've written S2N_ADD_IS_OVERFLOW_SAFE, but I suppose for an actual "overflow" check I should use SIZE_MAX. I'll just then need another check for <=SSIZE_MAX before I assign to total_size_out.

Why are these values signed ;_;